File: 15/15.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"sort"
	"strconv"
)

type Position struct {
	x int
	y int
}

func abs(x int) int {
	if x < 0 {
		return -x
	} else {
		return x
	}
}

func max(x, y int) int {
	if x > y {
		return x
	} else {
		return y
	}
}

func ManhattenDistance(a, b Position) int {
	return abs(a.x-b.x) + abs(a.y-b.y)
}

type SensorBeacon struct {
	sensor Position
	beacon Position
}

type Range struct {
	left  int
	right int
}

func NonOverlapping(ranges []Range) []Range {
	sort.Slice(ranges, func(i, j int) bool {
		return ranges[i].left < ranges[j].left
	})

	newRanges := []Range{}

	cur := Range{left: 0, right: -2}
	for _, r := range ranges {
		if cur.left <= r.left && r.left <= cur.right+1 {
			cur.right = max(cur.right, r.right)
		} else {
			if cur.left <= cur.right {
				newRanges = append(newRanges, cur)
				// fmt.Println(newRanges)
			}

			cur = r
		}
	}

	newRanges = append(newRanges, cur)
	// fmt.Println(newRanges)

	return newRanges
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)

	positionsRegex := regexp.MustCompile(`Sensor at x=([+-]?\d+), y=([+-]?\d+): closest beacon is at x=([+-]?\d+), y=([+-]?\d+)`)

	positions := []SensorBeacon{}

	for scanner.Scan() {
		line := scanner.Text()

		match := positionsRegex.FindStringSubmatch(line)

		if match != nil {
			vals := []int{}
			for _, s := range match[1:] {
				n, _ := strconv.Atoi(s)
				vals = append(vals, n)
			}

			sensor := Position{x: vals[0], y: vals[1]}
			beacon := Position{x: vals[2], y: vals[3]}

			positions = append(positions, SensorBeacon{sensor: sensor, beacon: beacon})

			// fmt.Printf("%v\n", positions)
		}
	}

	for y := 0; y <= 4000000; y += 1 {
		positionRanges := []Range{}

		for _, pos := range positions {
			sensor, beacon := pos.sensor, pos.beacon

			beaconDistance := ManhattenDistance(sensor, beacon)

			yDistance := abs(y - sensor.y)

			if yDistance <= beaconDistance {
				width := beaconDistance - yDistance

				left, right := sensor.x-width, sensor.x+width

				// # if beacon[1] == y:
				// #     if left == beacon[0]:
				// #         left += 1
				// #     if right == beacon[0]:
				// #         right -= 1

				if left <= right {
					posRange := Range{left: left, right: right}
					positionRanges = append(positionRanges, posRange)
				}
			}
		}

		nonOverlappingRanges := NonOverlapping(positionRanges)

		if y%400000 == 0 {
			fmt.Printf("%d%% done\n", 100*y/4000000)
		}

		if len(nonOverlappingRanges) >= 2 {
			// # y = 2766584 (-864133, 3135799) (3135801, 4102797)
			// # 3135800 * 4000000 + 2766584 = 12543202766584
			fmt.Printf("y = %2d %v\n", y, nonOverlappingRanges)
		}
	}
}

File: 15/15.py
import sys
import re


def manhatten_distance(pos1, pos2):
    return abs(pos1[0] - pos2[0]) + abs(pos1[1] - pos2[1])


def non_overlapping(ranges):
    left, right = 0, -2
    for l, r in sorted(ranges):
        # print(l, r)
        if left <= l <= right + 1:
            right = max(right, r)
        else:
            if left <= right:
                yield (left, right)

            left, right = l, r

    yield (left, right)


positions = []

for line in sys.stdin:
    if match := re.match(r'Sensor at x=([+-]?\d+), y=([+-]?\d+): closest beacon is at x=([+-]?\d+), y=([+-]?\d+)', line):
        sx, sy, bx, by = map(int, match.groups())

        sensor = (sx, sy)
        beacon = (bx, by)

        positions.append((sensor, beacon))


# for y in range(0, 21):
for y in range(0, 4000001):
    position_ranges = []

    for sensor, beacon in positions:
        beacon_distance = manhatten_distance(sensor, beacon)

        # print(sensor, beacon, abs(y - sy), distance)

        y_distance = abs(y - sensor[1])

        if y_distance <= beacon_distance:
            width = beacon_distance - y_distance

            left, right = sensor[0] - width, sensor[0] + width

            # if beacon[1] == y:
            #     if left == beacon[0]:
            #         left += 1
            #     if right == beacon[0]:
            #         right -= 1

            if left <= right:
                pos_range = (left, right)
                position_ranges.append(pos_range)

    ranges = list(non_overlapping(position_ranges))
    if len(ranges) >= 2:
        # y = 2766584 (-864133, 3135799) (3135801, 4102797)
        # 3135800 * 4000000 + 2766584 = 12543202766584
        print(f'y = {y:2}', *ranges)

    if y % 400000 == 0:
        print(f'{100 * y / 4000000}% done')

    # print(position_ranges)
    # print(sum(r - l + 1 for l, r in non_overlapping(position_ranges)))
